prow/cmd/deck/job_history.go

```diff
@@ -0,0 +1,389 @@
+}
+
+func getPath(bkt *storage.BucketHandle, root, id, fname string) (string, error) {
+	if strings.HasPrefix(root, logsPrefix) {
+		return path.Join(root, id, fname), nil
+	}
+	symLink := path.Join(root, id+".txt")
+	dir, err := resolveSymLink(bkt, symLink)
+	if err != nil {
+		return "", fmt.Errorf("failed to resolve sym link: %v", err)
+	}
+	return path.Join(dir, fname), nil
+}
+
+// reads specified JSON file in to `data`
+func readJSON(bkt *storage.BucketHandle, root, id, fname string, data interface{}) error {
+	p, err := getPath(bkt, root, id, fname)
+	if err != nil {
+		return fmt.Errorf("failed to get path: %v", err)
+	}
+	rawData, err := readObject(bkt.Object(p))
+	if err != nil {
+		return fmt.Errorf("failed to read %s for build %s: %v", fname, id, err)
+	}
+	err = json.Unmarshal(rawData, &data)
+	if err != nil {
+		return fmt.Errorf("failed to parse %s for build %s: %v", fname, id, err)
+	}
+	return nil
+}
+
+// Lists the GCS "directory names" immediately under prefix.
+func listSubDirs(bkt *storage.BucketHandle, prefix string) ([]string, error) {
+	if !strings.HasSuffix(prefix, "/") {
+		prefix += "/"
+	}
+	dirs := []string{}
+	it := bkt.Objects(context.Background(), &storage.Query{
+		Prefix:    prefix,
+		Delimiter: "/",
+	})
+	for {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return dirs, err
+		}
+		if attrs.Prefix != "" {
+			dirs = append(dirs, attrs.Prefix)
+		}
+	}
+	return dirs, nil
+}
+
+// Lists all GCS keys with given prefix.
+func listAll(bkt *storage.BucketHandle, prefix string) ([]string, error) {
+	keys := []string{}
+	it := bkt.Objects(context.Background(), &storage.Query{
+		Prefix: prefix,
+	})
+	for {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return keys, err
+		}
+		keys = append(keys, attrs.Name)
+	}
+	return keys, nil
+}
+
+// Gets all build ids for a job.
+func listBuildIDs(bkt *storage.BucketHandle, root string) ([]int64, error) {
+	ids := []int64{}
+	if strings.HasPrefix(root, logsPrefix) {
+		dirs, err := listSubDirs(bkt, root)
+		if err != nil {
+			return ids, fmt.Errorf("failed to list GCS directories: %v", err)
+		}
+		for _, dir := range dirs {
+			i, err := strconv.ParseInt(path.Base(dir), 10, 64)
+			if err == nil {
+				ids = append(ids, i)
+			}
+		}
+	} else {
+		keys, err := listAll(bkt, root)
+		if err != nil {
+			return ids, fmt.Errorf("failed to list GCS keys: %v", err)
+		}
+		for _, key := range keys {
+			matches := linkRe.FindStringSubmatch(key)
+			if len(matches) == 2 {
+				i, err := strconv.ParseInt(matches[1], 10, 64)
+				if err == nil {
+					ids = append(ids, i)
+				}
+			}
+		}
+	}
+	return ids, nil
+}
+
+func jobHistURL(url *url.URL) (string, string, int64, error) {
+	p := strings.TrimPrefix(url.Path, "/job-history/")
+	s := strings.SplitN(p, "/", 2)
+	if len(s) < 2 {
+		return "", "", emptyID, fmt.Errorf("invalid path: %v", url.Path)
```

Comment:
nit: Mention the expected format in the error message so that the error message is actionable.